DTS clean and fixups

mbolivar opened the pull request:
Some documentation fixes and cleanups related to DTS, which propagate out into the boards.

This is a prep series for some other work, but it's useful on its own, especially the documentation updates.

Commit:
doc: flash_partitions: rework

Re-work the flash partitions documentation:

- add sections to provide more complete context

- link to useful information on DT overlays elsewhere in the
  documentation

- touch up some phrasing

- inline a smaller, self-contained example within the page rather than
  linking to an entire board dts

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
doc: application: re-work DT overlay section

Edits and additions for clarity and correctness, based on the current
contents of the file (dts.cmake) which controls the DTS build.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
dts: common.dts: trivial whitespace cleanup

Trivial.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
boards: delete misleading comments in flash partitions

The space referred to in these comments was once truly reserved for
the application: no zephyr partition would ever use it.

Now, though, that's not the case, since NFFS will use these sectors
when enabled.

Just delete the comments.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

dbkinder commented on doc/application/application.rst:
```diff
@@ -966,50 +966,60 @@ preferred, since it correctly handles dependencies between options.
 
-#. Common overlays conditionally included by :file:`dts/common/common.dts`.
+#. The board's device tree source file, which by default is the Zephyr
+   file :file:`boards/ARCHITECTURE/BOARD/BOARD.dts`. (This location
```
Since it's not actually a folder called ARCHITECTURE or BOARD, it might be better to write this as:

    :file:`boards/<ARCHITECTURE>/<BOARD>/<BOARD>.dts`

dbkinder commented on doc/application/application.rst:
```diff
@@ -966,50 +966,60 @@ preferred, since it correctly handles dependencies between options.
+  takes precedence.
+
+- The environment variable :envvar:`DTC_OVERLAY_FILE` is checked
+  next. This mechanism is deprecated.
```
If it's deprecated, should it even be listed here? Or maybe write this:

    - In previous releases, the environment variable :envvar:`DTC_OVERLAY_FILE`
      was checked next, but this mechanism is now deprecated and not supported.

mbolivar replied:
I'm listing it here because despite being deprecated, the mechanism is still supported / still works. If users have this environment variable set, it will affect their build. For that reason I think it's worth documenting the behavior to avoid confusion. OK with you?

dbkinder replied:
So you're "deprecated" comment is saying it may not work in the future then.  I get it.

mbolivar replied:
Right. I'll add a note saying that users shouldn't rely on this, and ought to migrate to the previous mechanism in the list.

dbkinder commented on doc/devices/dts/flash_partitions.rst:
```diff
@@ -1,49 +1,228 @@
+
+You can define partitions for any flash device on the system.
+
+Most SoCs supported by Zephyr which have flash support in device tree
```
Too many whiches :)   How about:

    Most Zephyr-supported SoCs with flash support in device tree
    will define a label ``flash0``.   This label refers to the primary
    on-die flash programmed to run Zephyr. To generate partitions
    for this device, add the following snippet to a device tree overlay file:

dbkinder reviewed (changes requested):
looks pretty good... some suggested edits

carlescufi commented on doc/application/application.rst:
```diff
@@ -966,50 +966,60 @@ preferred, since it correctly handles dependencies between options.
-#. An application's current DT configuration (i.e. the file
-   :file:`zephyr/<board>.dts.pre.tmp` in the build directory).
+The Zephyr build system begins creation of a device tree by running
+the C preprocessor on a file which includes the following:
```
You've inverted the order here which makes it inconsistent with the equivlalent list describing the Kconfig and `.conf` processing. I would recommend we keep both in sync in terms of order?

carlescufi replied:
Do we really need to tell the user that we generate that file and then dtc is run on it? sounds like an implementation detail

mbolivar replied:
> You've inverted the order here which makes it inconsistent with the equivlalent list describing the Kconfig and .conf processing. I would recommend we keep both in sync in terms of order?

What do you mean? The new text is a description of how the DT is built; the old text is just incorrect AFAICT. What am I missing?

> Do we really need to tell the user that we generate that file and then dtc is run on it? sounds like an implementation detail

The reason I chose to include it is because it lets them know that they can use preprocessor symbols -- and in particular, which ones. That should help them write their overlays.

I can remove it if you prefer this to remain undocumented.

carlescufi commented on doc/application/application.rst:
```diff
@@ -966,50 +966,60 @@ preferred, since it correctly handles dependencies between options.
+describe the hardware it runs on. This section describes how you can
+modify an application build's device tree using overlay files.
 
-The Zephyr build system takes a Device Tree configuration
```
This text was aligned with the similar Kconfig section, now it is not, which makes it harder for the user to understand that they are similar in nature in terms of what the app can do with them?

mbolivar replied:
The build system's Kconfig and DT behaviors are different, so I don't think it's accurate to try to force them to look the same in documentation. What am I missing?

carlescufi replied:
Maybe I didn't explain myself properly. What I meant is that the Kconfig doc and the DTS doc listed the sources in reverse order of "overriding". So the first listed item's values can be overriden by later items, until the last one which cannot be overriden and hast the "last word".

mbolivar replied:
Your meaning is clear, but I believe the updated text is describing exactly that.

At cmake generation time, each device tree file is included into empty_file.c file in the order determined by the dts_files CMake variable here:

https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/dts.cmake#L34

The output of running the C pre-processor on that file is then unconditionally stored into BOARD.dts.pre.tmp. That overwrites the contents of the existing .dts.pre.tmp, rather than taking any configuration values from there, which the previous text incorrectly says it does. The macros in AUTOCONF_H are included before any device tree files in dts_files.

Scrolling up earlier in the file, dts_files contains the following, in order:

- DTS_SOURCE (which defaults to BOARD_ROOT/boards/ARCH/BOARD/BOARD.dts)
- DTS_COMMON_OVERLAYS (defaults to ZEPHYR_BASE/dts/common/common.dts)
- The contents of DTC_OVERLAY_FILE, as a list

This is the same order described in the documentation, in increasing order of ability to override. So I think this is correct and doing what you want. What am I missing?

carlescufi replied:
After reviewing this text and the `dts.cmake` file, I do believe you are correct. Thanks for the explanation.

carlescufi reviewed (changes requested):
Nice additions! I've added some comments to the DTS section

nvlsianpu commented:
apart of remarks by the other - this doc looks good. 
I'm reluctant to the commit of `boards: dts: delete now-unnecessary comments` - (prefer more doc/comments than less) but understand that you want to put some tidiness here, so I can beer this.

mbolivar commented:
@dbkinder comments addressed (made two changes, responded to your third review comment).

mbolivar commented:
> apart of remarks by the other - this doc looks good.
> I'm reluctant to the commit of boards: dts: delete now-unnecessary comments - (prefer more doc/comments than less) but understand that you want to put some tidiness here, so I can bear this.

How would you feel about replacing the comments with links to the documentation page?

nvlsianpu commented:
>how would you feel about replacing the comments with links to the documentation page?

@mbolivar This sounds good to me.

dbkinder reviewed (approved):
thanks.

Commit:
boards: dts: replace comments with links to documentation

Let's point users at the latest documentation on flash
partitions from each of the board files, rather than duplicating some
of the information in it.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

mbolivar commented:
@carlescufi @nvlsianpu updated
